### ISSUE TITLE
ACSF Config Management Approach

### ISSUE THREAD

**Issue by wouteradem** [acquia/blt#2276]
# Profile config strategy on ACSF
This is a possible enhancement for managing configuration into site profiles using ACSF.

## Use Case
As an example use case we can think of company X having N brands. A single brand can exist of a number of countries. And a country can consist of a number of regions. An example of such a (generic) profile tree structure can be found in the image below:

<img width="767" alt="screen shot 2017-11-14 at 11 31 25" src="https://user-images.githubusercontent.com/3283630/32775138-e1b365b0-c92e-11e7-89f5-fdd0ee0cd252.png">

## Current problem
It is difficult to swap out a profile on a local development environment on a multisite environment for ACSF. There is no proper way to organise configuration files.

## Approach

- BLT checkout profile

Use D8's default CMI location `config/default` as the default location for config files. Now, every site is mapped on a custom site profile (see generic profile tree structure above). Every profile contains a `config/sync` directory and an `.install` file.  Introduce three new BLT commands that allows one to check out a profile and copy all relevant config to `config/default` based on the profile tree structure.  

- BLT sieve profile

Takes the config/default directory and checks the current checked out profile and recursively loops over the generic profile structure config/sync directories (except for the current profile) and filters out the same config files (moves the files) and puts them one level higher (i.e. the base profile). This allows config files to be re-used over different profiles (i.e. avoid duplication)

- BLT get profile

Very simple command to get the current installed profile. (Can be extended to give a tree structure of the profiles for this codebase)

The structure of the commands are [TYPE]:[ACTION]:[CONTEXT] where
[TYPE] = acsf
[ACTION] = checkout, sieve or get
[CONTEXT] = profile

1. `blt acsf:checkout:profile [profile_name]`
2. `blt acsf:sieve:profile`
3. `blt acsf:get:profile`

Examples:

- Checkout myBe[LANGUAGE]Brand[N] where LANGUAGE = NL and N = 1
`blt acsf:checkout:profile myBeNlBrand1` would look into the `config/sync` directory of the custom profile _myBeNlBrand1_ and copy those `.yml` files to `config/default`. Next it would look into its parent (i.e. base profile) _myBeBrand1_ and copy those `.yml` files to `config/default`. Next it would look into its parent (i.e. base profile) _myInternationalBrand1_ and copy those` .yml` files to `config/default`. All config for that profile now lives in `config/default`.
Using a `drush cim` one can then import the config for that profile and finalise the checkout process.

- Allows you to work on generic profile feature requirements for example for _myBeBrand1_ , checkout _myBeBrand1_ profile and then build the feature and export to custom/profiles/myBeBrand1/config/sync by using drush cim --destination=custom/profiles/myBeBrand1/config/sync.

- When having an existing codebase you can use the sieve command to build up a generic config structure (given that you've build your profile tree structure).

## Status

1. `blt acsf:checkout:profile [profile_name]`

Done and tested.

2. `blt acsf:sieve:profile`

Done but still in test phase.

3. `blt acsf:get:profile`

Done and tested.

**Comment by lcatlett** [comment]
@wouteradem  I think this request might be re-creating functionality already available in config split and the default BLT split configuration. It may be worthwhile adding a "profile" split to BLT that is automatically generated based on the available profiles in docroot/profiles/custom and weight the splits accordingly for a site's active splits. I have successfully implemented this architecture with the current version of BLT and some tweaks to the pre-configured split directory paths and my config directory looks like this: 

└── config
    ├── default
    ├── envs
    │   ├── ci
    │   ├── dev
    │   ├── local
    │   ├── prod
    │   └── test
    ├── profile
    │   ├── profile1
    │   └── profile2
    └── site
        ├── site1
        ├── site2
        └── site3

I chose this approach because I was able to just use config split and keep all my config in one directory rather than within the docroot. Would this accomplish the functionality you are looking for?
